File: 1000s/1766.py
from collections import deque
from queue import PriorityQueue

n, m = map(int, input().split())

arr = [[] for i in range (n+1)]
indegree = [0 for i in range (n+1)]

for i in range(m):
    a, b = map(int, input().split())
    arr[a].append(b)
    indegree[b] += 1

pq = PriorityQueue()
for i in range(1, n+1):
    if indegree[i] == 0:
        pq.put(i);

while not pq.empty():
    top = pq.get()
    print(top, end=" ")

    for i in range(len(arr[top])):
        indegree[arr[top][i]] -= 1
        if indegree[arr[top][i]] == 0:
            pq.put(arr[top][i])

File: 1000s/1766.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <algorithm>

using namespace std;

int main()
{
    int n, m;
    cin >> n >> m;
    vector <vector <int>> arr(n+1);
    vector <int> indegree(n+1);
    for(int i = 0; i < m; i++)
    {
        int a, b;
        cin >> a >> b;
        arr[a].push_back(b);
        indegree[b]++;
    }

    priority_queue <int, vector <int>, greater<>> pq;
    for(int i = 1; i <= n; i++)
    {
        if(indegree[i] == 0)
        {
            pq.push(i);
        }
    }
    while(!pq.empty())
    {
        int top = pq.top();
        cout << top << " ";
        pq.pop();

        for(int i = 0; i < arr[top].size(); i++)
        {
            indegree[arr[top][i]]--;
            if(indegree[arr[top][i]] == 0)
            {
                pq.push(arr[top][i]);
            }
        }
    }
}